[quantization] Add Runtime flag for quantized backend.

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25680 [quantization] Add Runtime flag for quantized backend.**

Add a runtime flag to choose between FBGEMM and QNNPACK when compiled with both.

The flag can be set by using `torch.backends.quantized.engine = torch.fbgemm/torch.qnnpack` or `ctx::setPreferredQuantizedEngine(at::QEngine)`

Differential Revision: [D17198233](https://our.internmc.facebook.com/intern/diff/D17198233/)

supriyar committed:
[quantization] Add Runtime flag for quantized backend.

Add a runtime flag to choose between FBGEMM and QNNPACK when compiled with both.

The flag can be set by using `torch.backends.fbgemm/qnnpack` or `ctx::userEnabledFBGEMM/QNNPACK()`

Differential Revision: [D17198233](https://our.internmc.facebook.com/intern/diff/D17198233/)

ljk53 commented:
Adding @gchanan who might also be able to help review user facing API changes.

supriyar commented on torch/backends/fbgemm/__init__.py:
```diff
@@ -0,0 +1,29 @@
+
+# This is the sys.modules replacement trick, see
+# https://stackoverflow.com/questions/2447353/getattr-on-a-module/7668273#7668273
+sys.modules[__name__] = FbgemmModule(sys.modules[__name__], __name__)
```
I followed the convention used in Cudnn module https://github.com/pytorch/pytorch/blob/master/torch/backends/cudnn/__init__.py

jgong5 commented on aten/src/ATen/Context.h:
```diff
@@ -119,6 +123,8 @@ class CAFFE2_API Context {
   bool deterministic_cudnn = false;
   bool benchmark_cudnn = false;
+  bool enabled_fbgemm = true;
+  bool enabled_qnnpack = false;
```
How about we use one enum flag instead of multiple boolean flags like what @dzhulgakov proposed? Users can force the specific engine to use with this flag. When it is not set, pick the engine with some predefined rules, e.g. pick the optimal engine according to HW platform.

https://github.com/pytorch/pytorch/pull/25338#issuecomment-526475302
> have global runtime API for debugging, something like torch.backends.quantized.engine which is None by default, but can be set to FBGEMM or QNNPACK

supriyar committed:
Update on "[quantization] Add Runtime flag for quantized backend."


Add a runtime flag to choose between FBGEMM and QNNPACK when compiled with both.

The flag can be set by using `torch.backends.fbgemm/qnnpack` or `ctx::userEnabledFBGEMM/QNNPACK()`

Differential Revision: [D17198233](https://our.internmc.facebook.com/intern/diff/D17198233/)

supriyar committed:
Update on "[quantization] Add Runtime flag for quantized backend."


Add a runtime flag to choose between FBGEMM and QNNPACK when compiled with both.

The flag can be set by using `torch.backends.fbgemm/qnnpack` or `ctx::userEnabledFBGEMM/QNNPACK()`

Differential Revision: [D17198233](https://our.internmc.facebook.com/intern/diff/D17198233/)

supriyar committed:
Update on "[quantization] Add Runtime flag for quantized backend."


Add a runtime flag to choose between FBGEMM and QNNPACK when compiled with both.

The flag can be set by using `torch.backends.fbgemm/qnnpack` or `ctx::userEnabledFBGEMM/QNNPACK()`

Differential Revision: [D17198233](https://our.internmc.facebook.com/intern/diff/D17198233/)

supriyar committed:
Update on "[quantization] Add Runtime flag for quantized backend."


Add a runtime flag to choose between FBGEMM and QNNPACK when compiled with both.

The flag can be set by using `torch.backends.fbgemm/qnnpack` or `ctx::userEnabledFBGEMM/QNNPACK()`

Differential Revision: [D17198233](https://our.internmc.facebook.com/intern/diff/D17198233/)

ljk53 reviewed (commented):
Looks like most of the code change follows qschema pattern.
Just one question about how should we name this new concept, qbackend? - I'm not sure whether people working on quantization have strong opinion, cc: @raghuramank100 @jerryzh168 @jamesr66a

supriyar commented on aten/src/ATen/Context.h:
```diff
@@ -119,6 +122,7 @@ class CAFFE2_API Context {
   bool enabled_cudnn = true;
   bool deterministic_cudnn = false;
   bool benchmark_cudnn = false;
+  at::QBackend quantized_engine = at::QBackend::FBGEMM;
```
c10/core/Backend already has a backend for quantization,  `Backend::QuantizedCPU`. I think this affects dispatch of quantization related stuff, etc so we don't want to modify this. 
I added `QBackend` to be more of a quantization specific parameter. I agree with renaming `quantized_engine` to `quantized_backend` to make it more consistent (or vice versa - calling it QEngine in case the word backend gets too confusing and overloaded).

jerryzh168 replied:
I think engine might be better, qbackend will be conflicting with current backend name, we want to reduce confusions. 
I think we can do
```
at::QEngine qengine = at::QEngine::FBGEMM;
```
or simply:
```
at::Engine engine = at::Engine::FBGEMM;
```

supriyar replied:
Updated to use QEngine instead of QBackend

supriyar committed:
Update on "[quantization] Add Runtime flag for quantized backend."


Add a runtime flag to choose between FBGEMM and QNNPACK when compiled with both.

The flag can be set by using `torch.backends.fbgemm/qnnpack` or `ctx::userEnabledFBGEMM/QNNPACK()`

Differential Revision: [D17198233](https://our.internmc.facebook.com/intern/diff/D17198233/)

supriyar committed:
Update on "[quantization] Add Runtime flag for quantized backend."


Add a runtime flag to choose between FBGEMM and QNNPACK when compiled with both.

The flag can be set by using `torch.backends.quantized.engine = torch.fbgemm/torch.qnnpack` or `ctx::setPreferredQuantizedEngine(at::QEngine)`

Differential Revision: [D17198233](https://our.internmc.facebook.com/intern/diff/D17198233/)

ljk53 commented on torch/csrc/QEngine.cpp:
```diff
@@ -0,0 +1,79 @@
+  return THPUtils_packString("torch." + name);
+}
+
+PyTypeObject THPQEngineType = {
```
@dzhulgakov mentioned offline that there is better way to do this type of binding. But it can be improved in separate diff.

ljk53 reviewed (approved):
LGTM

supriyar committed:
Update on "[quantization] Add Runtime flag for quantized backend."


Add a runtime flag to choose between FBGEMM and QNNPACK when compiled with both.

The flag can be set by using `torch.backends.quantized.engine = torch.fbgemm/torch.qnnpack` or `ctx::setPreferredQuantizedEngine(at::QEngine)`

Differential Revision: [D17198233](https://our.internmc.facebook.com/intern/diff/D17198233/)

dzhulgakov reviewed (commented):
One question about logic - is it really "preferred" engine or "required" engine? Not sure which one would be more convenient.

If it's "required", I'd expect "none" option to be default and setting the engine if it's not available to just throw.

But given that it's going to be used only in perf debugging tools, I think it's fine either way

gchanan commented on test/test_quantized.py:
```diff
@@ -913,7 +913,7 @@ def test_qlinear(self, batch_size, input_channels, output_channels, use_bias,
     def test_qlinear_unpack(self, W, use_channelwise):
         W, (W_scale, W_zp, torch_type) = W
-
+        torch.backends.quantized.engine = torch.fbgemm
```
ya for something like this, I'd probably just use a string (you can translate it to an integer at the python level so the C++ API is nicer).  See e.g. `reduction` here, just ignore the magic for setting a default, I don't think you need that.
https://github.com/pytorch/pytorch/blob/ffbffb69c6bf75e1a998844c8ca5b4a0fa90d4cc/aten/src/ATen/native/native_functions.yaml#L439
